### PR TITLE
Add query methods that return Slice to FirestoreTemplate to allow pagination

### LINKED ISSUE #2535: DatastoreTemplate DatastoreResultsIterable Cursor should be null when there is no next page
I am trying to implement cursor-based pagination and found that unable to determine whether DatastoreResultsIterable has the next page. Due to the DatastoreResultsIterable Cursor has NotNull value even there is no next page.

Given:
Datastore has some data in KIND
pageLimit > rowsCount for KIND

```
EntityQuery.Builder queryBuilder = Query.newEntityQueryBuilder()
				.setKind(KIND)
				.setOrderBy(StructuredQuery.OrderBy.desc(KEY))
				.setLimit(pageLimit);
```
```
if (urlSafeCursor != null) {
    queryBuilder.setStartCursor(Cursor.fromUrlSafe(urlSafeCursor));
}
```

When:
`DatastoreResultsIterable<DO> resultList = datastoreTemplate.query(queryBuilder.build(), DO.class);`

Expected:
resultList.getCursor() == null

Actual:
resultList.getCursor() != null

### PR DESCRIPTION
**Pull request by dmitry-s**
fixes #2535

### CODE REVIEW

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2019-2019 the original author or authors.
```
Comment (dzou) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java:2]: Copyright 2017-2020

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+import com.google.cloud.datastore.Cursor;
+
+/**
+ * An iterable that contains a cursor for the next page and can be used to determine
+ * if the next page exists.
+ *
+ * @author Dmitry Solomakha
+ */
+public class DatastoreNextPageAwareResultsIterable<T> implements Iterable<T> {
```
Comment (dzou) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java:29]: I don't think this is quite the right approach; it seems a bit heavyweight to create a new class to hold a boolean.

Also the only method that returns this is `nextPageAwareQuery`; however, other query methods such as `query`, `queryIterable`, and `queryByExample` also return `DatastoreResultsIterable` but are not able to take advantage of the next-page-aware feature.

I would try to get `DatastoreResultsIterable` to have access to the `hasNextPage` somehow.

I would suggest a couple alternatives:

1. Is it possible to add a method to DatastoreTemplate like `hasNextPage(Cursor c)` or `hasNextPage(DatastoreResultsIterable i)`? if so, the user can manually check for next page through usage of template.

2. If the query is cheap, you might as well pay the cost of `nextPageExists(..)` for each DatastoreResultsIterable; and build the `hasNextPage` parameter in there. no need for premature optimization

3. Add a method called `hasNextPage()` on the `DatastoreResultsIterable` which will run the query of checking if the cursor has a next page. This would require the Iterable obj have reference to the datastore client I suppose.

Comment (meltsufin) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java:29]: +1
Can we just extend `DatastoreResultsIterable`?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java [hunk 3]
```diff
@@ -276,6 +277,21 @@ public long count(Class<?> entityClass) {
...
 	}
 
+	@Override
+	public <T> DatastoreNextPageAwareResultsIterable<?> nextPageAwareQuery(StructuredQuery query, Class<T> entityClass) {
+		DatastoreResultsIterable resultsIterable = queryKeysOrEntities(query, entityClass);
+
+		return new DatastoreNextPageAwareResultsIterable(resultsIterable, nextPageExists(query, resultsIterable.getCursor()));
+	}
+
+	private boolean nextPageExists(StructuredQuery query, Cursor cursorAfter) {
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java:287]: Can we just move this code to `hasNextPage` in the `Iterable`?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java [hunk 4]
```diff
@@ -248,21 +248,19 @@ private Slice executeSliceQuery(Object[] parameters) {
...
+				StreamSupport.stream(results.spliterator(), false).collect(Collectors.toList());
 
 		return (Slice) this.processRawObjectForProjection(
-				new SliceImpl(result, pageable, nextResult.iterator().hasNext()));
+				new SliceImpl(result, getPageable(parameters, results.getCursor()), results.hasNextPage()));
+	}
+
+	private Pageable getPageable(Object[] parameters, Cursor cursor) {
+		ParameterAccessor paramAccessor = new ParametersParameterAccessor(getQueryMethod().getParameters(), parameters);
+		Pageable pageable = DatastorePageable.from(paramAccessor.getPageable(), cursor, null);
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java:262]: You can just return here without assigning to `pageable`.

**Review by meltsufin — CHANGES_REQUESTED** [review]

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreResultsIterable.java [hunk 5]
```diff
@@ -50,4 +61,20 @@ public Cursor getCursor() {
 	public Iterable<T> getIterable() {
 		return this.iterable;
 	}
+
+	public Optional<Boolean> getHasNextPage() {
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreResultsIterable.java:65]: why not just
`public boolean hasNextPage()`?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreResultsIterable.java [hunk 6]
```diff
@@ -50,4 +61,20 @@ public Cursor getCursor() {
...
+		return hasNextPage.get();
+	}
+
+	public void setHasNextPageQuery(Supplier<Boolean> query) {
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreResultsIterable.java:69]: I don't think a `public` setter makes sense here. Can you just provide all the necessary objects to determine `hasNextPage` into the constructor?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java [hunk 7]
```diff
@@ -272,10 +273,25 @@ public long count(Class<?> entityClass) {
 			resultsIterable = new DatastoreResultsIterable<>(convertEntitiesForRead(results, entityClass),
 					results.getCursorAfter());
 		}
+		if (query instanceof StructuredQuery) {
+			resultsIterable.setHasNextPageQuery(
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java:277]: Instead of calling `setHasNextPageQuery` can this block of code just move to `DatastoreResultsIterable`?
Also, why does this only apply to `StructuredQuery`?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java (hunk 3 above)
Comment (meltsufin) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java:287]: Same question as before.

**Review by meltsufin — CHANGES_REQUESTED** [review]

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreResultsIterable.java (hunk 5 above)
Comment (dmitry-s) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreResultsIterable.java:65]: @meltsufin Because it is not always there. We can only compute it for `StructuredQuery` queries, and we can't compute it for GQL queries. So we need to represent 3 values - true, false, empty.

Comment (meltsufin) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreResultsIterable.java:65]: @dmitry-s I must be missing something here. Both, `GqlQuery` and `StructuredQuery`, support cursors. So, I would image that it's possible to try to retrieve the next item using the cursor to determine `hasNext()`. How are they different in this respect?
Also, if there is a fundamental difference, then we should probably have two different classes for their results. So, maybe extend `DatastoreResultsIterable` and add `hasNextPage()`?

Comment (dmitry-s) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreResultsIterable.java:65]: The problem with `GqlQuery` is we need to set limit to 1 to avoid fetching the whole next page, and `GqlQuery` doesn't allow to change that.

Comment (meltsufin) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreResultsIterable.java:65]: So, for `GqlQuery` we would need to fetch the whole page? Is there a problem with that other than the potential performance overhead?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+import com.google.cloud.datastore.Cursor;
+import com.google.cloud.datastore.StructuredQuery;
+
+import org.springframework.data.util.Lazy;
+
+public class DatastoreNextPageAwareResultsIterable<T> implements DatastoreResultsIterable<T> {
+	private DatastoreResultsIterable datastoreResultsIterable;
+
+	private Cursor cursor;
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java:29]: Never used.

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+package org.springframework.cloud.gcp.data.datastore.core;
+
+import java.util.Iterator;
+
+import com.google.cloud.datastore.Cursor;
+import com.google.cloud.datastore.StructuredQuery;
+
+import org.springframework.data.util.Lazy;
+
+public class DatastoreNextPageAwareResultsIterable<T> implements DatastoreResultsIterable<T> {
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java:26]: `DatastoreStructuredQueryResultsIterable` maybe?

Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java:26]: Please add class-level javadoc and `@since 1.2.6`.

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	private DatastoreResultsIterable datastoreResultsIterable;
+
+	private Cursor cursor;
+
+	private StructuredQuery structuredQuery;
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java:31]: Never used.

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java [hunk 11]
```diff
@@ -266,16 +267,28 @@ public long count(Class<?> entityClass) {
...
 		}
+		if (query instanceof StructuredQuery) {
+			resultsIterable = new DatastoreNextPageAwareResultsIterable(
+					resultsIterable, results.getCursorAfter(), (StructuredQuery) query, this);
+		}
 		maybeEmitEvent(new AfterQueryEvent(resultsIterable, query));
 		return resultsIterable;
 	}
 
+	public boolean nextPageExists(StructuredQuery query, Cursor cursorAfter) {
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java:284]: Javadoc missing

`queryHasNextPage` maybe?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+import java.util.Iterator;
+
+import com.google.cloud.datastore.Cursor;
+import com.google.cloud.datastore.StructuredQuery;
+
+import org.springframework.data.util.Lazy;
+
+public class DatastoreNextPageAwareResultsIterable<T> implements DatastoreResultsIterable<T> {
+	private DatastoreResultsIterable datastoreResultsIterable;
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreNextPageAwareResultsIterable.java:27]: I don't think you're gaining much by delegating to another iterable. I would just use the currently unused instance variables instead.

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreSimpleResultsIterable.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.datastore.core;
+
+import java.util.Iterator;
+
+import com.google.cloud.datastore.Cursor;
+
+public class DatastoreSimpleResultsIterable<T> implements DatastoreResultsIterable<T> {
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreSimpleResultsIterable.java:23]: javadoc missing

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java [hunk 14]
```diff
@@ -245,24 +245,25 @@ private String mapToFieldName(PropertyDescriptor propertyDescriptor) {
...
 
 	private Slice executeSliceQuery(Object[] parameters) {
+		DatastoreNextPageAwareResultsIterable<?> results =
+				(DatastoreNextPageAwareResultsIterable<?>)
+						this.datastoreOperations.queryKeysOrEntities(buildSliceQuey(parameters), this.entityType);
+
+		return (Slice) this.processRawObjectForProjection(
+				new SliceImpl(results.toList(),
+						getPageable(parameters, results.getCursor()),
+						results.hasNextPage()));
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java:255]: Does it make sense to make `SliceImpl` also evaluate `hasNext()` lazily?

**Review by meltsufin — CHANGES_REQUESTED** [review]

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java (hunk 14 above)
Comment (dmitry-s) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java:255]: It is a Spring Data class and expects a boolean parameter.

Comment (meltsufin) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java:255]: Can we extend `SliceImpl`?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java [hunk 15]
```diff
@@ -272,11 +275,31 @@
 	/**
 	 * Finds objects by using a Cloud Datastore query. If the query is a key-query, then keys are
 	 * returned.
+	 * Resulting Slice can be used to get a Pageable for the next page or to determine if next page exists.
 	 * @param query the query to execute.
 	 * @param entityClass the type of object to retrieve.
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java:280]: Can you clarify which type to specify for key?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java [hunk 16]
```diff
@@ -272,11 +275,31 @@
 	/**
 	 * Finds objects by using a Cloud Datastore query. If the query is a key-query, then keys are
 	 * returned.
+	 * Resulting Slice can be used to get a Pageable for the next page or to determine if next page exists.
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java:278]:
```suggestion
	 * Resulting Slice can be used to get a Pageable for the next page or to determine if next the page exists.
```

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java [hunk 17]
```diff
@@ -272,11 +275,31 @@
...
+	<T> Slice<?> queryKeysOrEntitiesSlice(StructuredQuery query, Class<T> entityClass, Pageable pageable);
+
+	/**
+	 * Runs a query that checks if there is at least one entry after the cursor.
+	 *
+	 * @param query the query that returned the cursor
+	 * @param cursorAfter the cursor
+	 * @return true if next page exists for the given query and cursor
+	 */
+	boolean nextPageExists(StructuredQuery query, Cursor cursorAfter);
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java:294]: Do we still need this method, since `Slice` has `hasNextPage()`?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java [hunk 18]
```diff
@@ -261,6 +266,32 @@ public long count(Class<?> entityClass) {
 				: null;
 	}
 
+	@Override
+	public <T> Slice<?> queryKeysOrEntitiesSlice(StructuredQuery query, Class<T> entityClass, Pageable pageable) {
+		DatastoreResultsIterable<?> results = queryKeysOrEntities(applyPageable(query, pageable), entityClass);
+		return new SliceImpl<>(results.toList(),
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java:272]: Can we extend `SliceImpl` such that `hasNextPage()` is lazy?
Feel free to leave for another PR though.

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java [hunk 19]
```diff
@@ -272,11 +275,31 @@
...
+	 * Resulting Slice can be used to get a Pageable for the next page or to determine if next page exists.
 	 * @param query the query to execute.
 	 * @param entityClass the type of object to retrieve.
+	 * @param pageable that indicates page number and page size
 	 * @param <T> the type of object to retrieve.
-	 * @return a list of the objects found. If no keys could be found the list will be
-	 * empty.
+	 * @return a Slice containing found objects
+	 */
+	<T> Slice<?> queryKeysOrEntitiesSlice(StructuredQuery query, Class<T> entityClass, Pageable pageable);
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java:285]: Can you remind me, what is the benefit of combining `queryEntities()` and `queryKeys()` into one method?
Sonar is identifying this as a "code smell".

**Review by meltsufin — CHANGES_REQUESTED** [review]

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java (hunk 18 above)
Comment (dmitry-s) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java:272]: I don't think it is necessary because if you don't need  `hasNextPage()`, you could just use the `queryKeysOrEntities` method instead.

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java (hunk 19 above)
Comment (dmitry-s) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java:285]: The code would be identical for these cases, so we are avoiding duplication this way.

Comment (dmitry-s) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java:285]: Let me see if I can refactor it to avoid duplication.

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java (hunk 18 above)
Comment (meltsufin) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java:286]: What about the use in repositories?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java [hunk 20]
```diff
@@ -267,8 +268,21 @@ public long count(Class<?> entityClass) {
...
-	public <T> Slice<?> queryKeysOrEntitiesSlice(StructuredQuery query, Class<T> entityClass, Pageable pageable) {
-		DatastoreResultsIterable<?> results = queryKeysOrEntities(applyPageable(query, pageable), entityClass);
+	public <T> Slice<Key> queryKeysSlice(KeyQuery query, Class<T> entityClass, Pageable pageable) {
+		return buildSlice(query, pageable, Key.class);
+	}
+
+	@Override
+	public <T> Slice<T> queryEntitiesSlice(StructuredQuery query, Class<T> entityClass, Pageable pageable) {
+		if (query instanceof KeyQuery) {
+			throw new DatastoreDataException("query must be an EntityQuery or a ProjectionEntityQuery");
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java:278]: Why don't you check for those types in the condition then? What if other query types are added in the future?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java [hunk 21]
```diff
@@ -269,14 +272,35 @@
...
 	 * Finds objects by using a Cloud Datastore query. If the query is a key-query, then keys are
 	 * returned.
 	 * @param query the query to execute.
 	 * @param entityClass the type of object to retrieve.
 	 * @param <T> the type of object to retrieve.
-	 * @return a list of the objects found. If no keys could be found the list will be
-	 * empty.
+	 * @return an iterable containing found objects and a cursor
 	 */
 	<T> DatastoreResultsIterable<?> queryKeysOrEntities(Query query, Class<T> entityClass);
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java:305]: Can we also deprecate this method and bifurcate it like you did for `Slice`? This can be in a separate PR.

**Review by meltsufin — CHANGES_REQUESTED** [review]

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java (hunk 21 above)
Comment (dmitry-s) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreOperations.java:305]: created an issue: #2559

**Review by meltsufin — APPROVED** [review]

**Comment by meltsufin** [comment]
@dmitry-s Did you mean `DatastoreTemplate`? 😄
